Week 1 notes

krishnakumarg1984 opened the pull request.

mmcleod89 commented on 01projects/sec01Git.md:
````diff
@@ -0,0 +1,279 @@
+
+``` sh
+$ git config --global user.name "Full name"
+$ git config --global user.name "email address"
````
```suggestion
$ git config --global user.email "email address"
```

mmcleod89 commented on 01projects/sec01Git.md:
````diff
@@ -0,0 +1,279 @@
+$ git push -u origin main
+```
+
+The `git push` command facilities pushing of local changes to the remote. However, we need to specify the local referenc name for the remote repository (i.e. `origin` by convention as we set it up earlier) and the current branch (`main` in newer versions of Git). Note that we do not cover Git branches in depth in this introductory lesson. The `-u` flag indicates that the specified local branch (` ...
````
```suggestion
The `git push` command facilities pushing of local changes to the remote. However, we need to specify the local reference name for the remote repository (i.e. `origin` by convention as we set it up earlier) and the current branch (`main` in newer versions of Git). Note that we do not cover Git branches in depth in this introductory lesson. The `-u` flag indicates that the specified local branch (`main`) shall always be considered to correspond to the same branch in the specified remote. From now on, further commits in the local repository can be pushed by simply invoking `git push` without additional arguments.
```

mmcleod89 commented on 01projects/sec02SoftwareBuilds.md:
```diff
@@ -0,0 +1,86 @@
+
+### Multiple types of executable outputs
+
+While this above methodology suffices for small projects and enables to focus on learning and developing a solid understanding of the programming language itself, this approach does not scale well to meet the needs of larger, more complex projects.
```
```suggestion
While this above methodology suffices for small projects and enables one to focus on learning and developing a solid understanding of the programming language itself, this approach does not scale well to meet the needs of larger, more complex projects.
```

mmcleod89 commented on 01projects/sec02SoftwareBuilds.md:
```diff
@@ -0,0 +1,86 @@
+For instance on Windows, with the Microsoft C++ compiler and toolchain (MSVC), the build system is known as Visual Studio (an IDE exclusive to Windows) solutions, and on macOSes, the toolchain provided by Apple emits Xcode (an IDE exclusive to macOS) compatible output.
+
+In late 1999/early 2000, there was no compatible build system that worked across all Operating Systems, and developers had to go through the frustrating process of manually invoking multiple toolchains on different systems.
+It was during this time that, Kitware Inc, a company that was part of a consortium tasked with designing the C++-based Imaging Toolkit (itk) funded by a US national research grant, developed CMake, a cross-platform *build system generator* (a system that emits output for consumption by native build systems, which in-turn have to be invoked to produce the final build output).
```
The word "emits" might be a bit confusing here, I would maybe use "generates" or "outputs" and then say what the output actually is e.g. make files. Otherwise I think it won't be entirely clear what CMake is actually doing to someone who has never heard of it before.

krishnakumarg1984 replied:
@mmcleod89 Thank you. Have clarified this.

mmcleod89 commented on 01projects/sec03CMakeBackground.md:
```diff
@@ -0,0 +1,131 @@
+
+### Hierarchical project structure
+
+Large scale research projects typically use a hierarchical project structure using modular code patterns for [separation of concerns](https://en.wikipedia.org/wiki/Separation_of_concerns).
```
The phrase "modular code patterns" will definitely require some explanation as even I am not 100% sure what you mean by this

mmcleod89 commented on 01projects/sec04CMakeHelloWorld.md:
```diff
@@ -0,0 +1,237 @@
+The VERSION field is optional, but is conventional to populate it.
+Next, we specify the programming languages that the project uses.
+The keyword for C++ projects is `CXX`.
+It is helpful to specify the list of languages used in the project, so that cmake does not spend time introspecting the system for the presence of the toolchains for every supported language (CUDA, fortran, C, Java etc).
```
```suggestion
It is helpful to specify the list of languages used in the project, so that cmake does not spend time inspecting the system for the presence of the toolchains for every supported language (CUDA, fortran, C, Java etc).
```

krishnakumarg1984 replied:
@mmcleod89 It is actually called system introspection in CMake terminology (official online docs as well as in multiple textbooks that I referred to during my preparation). e.g. see here: https://cmake.org/cmake/help/latest/guide/tutorial/Adding%20System%20Introspection.html

mmcleod89 replied:
This is an interesting point. I'm happy to leave it up to you but here are my thoughts:
- A couple of comments below I think it makes sense to keep the term "System Introspection" (ln 64), especially given that this is the term that CMake is using. This is okay in this context because it just means the system inspecting itself, which may be true depending on how broadly we are defining "system" and what exactly we mean by this. 
- Grammatically, it is incorrect to say "instrospecting something" as in "introspecting the system". Introspecting cannot take an object, because it always means something inspecting itself. (One can say that "X is introspecting" but not "X is introspecting Y".) Since we are talking about one thing (cmake) examining another ("the system" i.e. environment, OS, hardware etc. - in this context it is clearly something separate from cmake) then "inspecting" is the correct verb. 
- I would argue that the same is true for the comment below (ln 53). "Hardware introspection" doesn't make sense to me because the hardware does not inspect itself.

krishnakumarg1984 replied:
@mmcleod89 Thank you. I have now used inspecting/inspection consistently throughout.

mmcleod89 commented on 01projects/sec02SoftwareBuilds.md:
```diff
@@ -0,0 +1,86 @@
+The most common form of distributed memory parallelism in high performance computing environments is a library written in plain C.
+This heterogeneous nature of research projects imply the correct invocation of multiple toolchains to produce the final executable output.
+
+The need to invoke multiple toolchains, conditionally enabling various capabilities, that are dependent on hardware introspection, for each platform and OS quickly becomes non-scalable.
```
```suggestion
The need to invoke multiple toolchains, conditionally enabling various capabilities, that are dependent on hardware inspection, for each platform and OS quickly becomes non-scalable.
```

mmcleod89 commented on 01projects/sec02SoftwareBuilds.md:
```diff
@@ -0,0 +1,86 @@
+The complexities of build tools have been acknowledged since nearly half a century, and several build automation tools (also known as build systems) have been devised over time to tackle the problem.
+An early such tool, which is still popular today, is [`make`](https://en.wikipedia.org/wiki/Make_(software)).
+In this build system, the library author writes a set of plain-text files, known as MakeFiles, and provides them along with the C++ source code.
+For system introspection, a set of companion tools called [GNU Autotools](https://en.wikipedia.org/wiki/GNU_Autotools) was developed to configure the build capabilities depending on the target processor's capabilities.
```
```suggestion
For system inspection, a set of companion tools called [GNU Autotools](https://en.wikipedia.org/wiki/GNU_Autotools) was developed to configure the build capabilities depending on the target processor's capabilities.
```

mmcleod89 commented on 01projects/sec04CMakeHelloWorld.md:
```diff
@@ -0,0 +1,237 @@
+
+The behaviour of targets, their features and their various properties can be set in three ways:
+- only on/for/affecting itself (using the PRIVATE keyword)
+- only on/for/affecting other targets that consume this current target (using the INTERFACE keyword)
+- on/for/affecting itself as well as other targets that consume this current target (using the INTERFACE keyword)
```
I think the last bullet point should be "PUBLIC", not interface

krishnakumarg1984 replied:
Fixed

mmcleod89 commented on 01projects/sec04CMakeHelloWorld.md:
```diff
@@ -0,0 +1,237 @@
+
+The following matrix may help slightly, and we may choose to refer to this at times throughout the course:
+
+Keyword selection matrix (Who needs/uses a dependency?)
+
+|                | Other (consuming targets) |
+|----------------|------------|---------|
+| **Current target** | **YES**        | **NO**      |
+| **YES**            | PUBLIC     | PRIVATE |
+| **NO**             | INTERFACE  | N/A     |
```
I would recommend calling it a "Table" rather than a "Matrix". 

I'm not sure if it's my browser viewing the md file but this table is not rendering for me... it might be okay on the website build though.

krishnakumarg1984 replied:
@mmcleod89 Fixed. Calling it a "table" now. The rendering issue seems to be fixed on the jekyll rendered pages

krishnakumarg1984 commented:
@mmcleod89 Thanks a lot for reviewing this. I have incorporated your feedback, and updated the text accordingly.

 I don't quite understand why the CI is failing (which blocks merging). Can someone please look into it since we need to release these notes to students soon! @dpshelio Could you please take a quick look?

JamieJQuinn commented:
@krishnakumarg1984 see [my changes](https://github.com/UCL/research-computing-with-cpp/pull/111/files#diff-28802fbf11c83a2eee09623fb192785e7ca92a3f40602a517c011b947a1822d3) to `config.yml`. Just implement them in this PR and we'll get on with life.

krishnakumarg1984 commented:
Let's ignore this PR. Unable to get CI to build this. Let's instead use #112
